webapp/src/components/surveyPost/surveyPost.tsx

```diff
@@ -0,0 +1,243 @@
+            newDraftResponse.responses[questionID] = response;
+            draftResponse.current = newDraftResponse;
+
+            // if this is the system rating question, submit response ASAP
+            if (questionID === linearScaleQuestionID.current) {
+                submitRating();
+            }
+        },
+        500,
+    );
+
+    const submitSurveyHandler = useCallback(async () => {
+        if (!draftResponse.current ||
+            Object.keys(draftResponse.current?.responses).length === 0
+        ) {
+            return;
+        }
+
+        const response = await submitSurveyResponse();
+        if (response.success && survey !== undefined) {
+            const newSurvey: UserSurvey = {
+                ...survey,
+                response: draftResponse.current,
+            };
+            setSurvey(newSurvey);
+            setErrorMessage('');
+        } else if (response.error) {
+            setErrorMessage('Failed to submit survey response. Please try again.');
+        }
+    }, [survey]);
+
+    const submitSurveyResponse = async () => {
+        // make API call here. Send draftResponse.current as payload...
+        return {success: true, error: false};
+    };
+
+    // this function is to submit the linear scale rating as soon as a user selects it,
+    // even without pressing the submit button.
+    const submitRating = async () => {
+        if (!linearScaleQuestionID.current || !draftResponse.current) {
+            return;
+        }
+
+        const ratingQuestionResponse = draftResponse.current?.responses[linearScaleQuestionID.current];
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        const payload: SurveyResponse = {
+            responses: {
+                [linearScaleQuestionID.current]: ratingQuestionResponse,
+            },
+            dateCreated: draftResponse.current?.dateCreated,
+        };
+
+        // send payload to submit survey response API here...
+    };
+
+    const renderedMessage = useMemo(() => {
+        // @ts-expect-error window is definitely MattermostWindow in plugins
+        const mmWindow = window as MattermostWindow;
+        const htmlString = mmWindow.PostUtils.formatText(post.message, {markdown: true});
+        return mmWindow.PostUtils.messageHtmlToComponent(htmlString, isRHS);
+    }, [post.message, isRHS]);
+
+    const renderQuestions = useMemo(() => {
+        if (!survey) {
+            return null;
+        }
+
+        return survey.questions.map((question) => {
+            let questionComponent: React.ReactNode;
+
+            switch (question.type) {
+            case 'linear_scale':
+                questionComponent = (
+                    <LinearScaleQuestion
+                        question={question}
+                        responseChangeHandler={questionResponseChangeHandler}
+                        disabled={disabled}
+                        value={survey.response?.responses[question.id] as number}
```

Comment:
Are all of these props expected to be the same for any type of question? If each question component has the same props, you could rewrite this switch statement using an object to map between the `type` and the corresponding component. I don't know how TS will handle this, but we've used this pattern elsewhere
```js
const questionComponents = {
    linear_scale: LinearScaleQuestion,
    text: TextQuestion,
};

const Question = questionComponents[question.type];

return (
    <Question
        question={question}
        responseChangeHandler={questionResponseChangeHandler}
        disabled={disabled}
        value={survey.response?.responses[question.id} // This is the line I don't know about
    />
);
```

Reply:
Nice idea! It worked! Done